File: jvm/src/test/kotlin/com/jetbrains/signatureverifier/tests/serialization/SignatureRecreationTests.kt
package com.jetbrains.signatureverifier.tests.serialization

import com.jetbrains.signatureverifier.PeFile
import com.jetbrains.signatureverifier.cf.MsiFile
import com.jetbrains.signatureverifier.crypt.SignedMessage
import com.jetbrains.signatureverifier.macho.MachoArch
import com.jetbrains.signatureverifier.serialization.Asn1PrimitiveSerializer
import com.jetbrains.signatureverifier.serialization.compareBytes
import com.jetbrains.signatureverifier.serialization.toContentInfo
import com.jetbrains.util.TestUtil
import com.jetbrains.util.TestUtil.getTestByteChannel
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.modules.SerializersModule
import kotlinx.serialization.modules.polymorphic
import org.bouncycastle.asn1.ASN1Primitive
import org.bouncycastle.asn1.ASN1Sequence
import org.bouncycastle.asn1.cms.SignedData
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource
import java.nio.file.Files
import java.nio.file.StandardOpenOption
import java.util.stream.Stream

class SignatureRecreationTests {

  @ParameterizedTest
  @MethodSource("SignedPEProvider")
  fun PE_Test(signedPeResourceName: String) {
    getTestByteChannel("pe", signedPeResourceName).use {
      val peFile = PeFile(it)
      val signatureData = peFile.GetSignatureData()
      val signedMessage = SignedMessage.CreateInstance(signatureData)

      RecreateSignatureTest(signedMessage, peFile.GetSignatureData().CmsData!!)
    }
  }

  @ParameterizedTest
  @MethodSource("SignedMachoProvider")
  fun Macho_Test(signedResourceName: String) {
    val machoFiles =
      Files.newByteChannel(
        TestUtil.getTestDataFile(
          "mach-o",
          signedResourceName
        ), StandardOpenOption.READ
      ).use {
        MachoArch(it).Extract()
      }

    for (machoFile in machoFiles) {
      val signatureData = machoFile.GetSignatureData()
      val signedMessage = SignedMessage.CreateInstance(signatureData)
      RecreateSignatureTest(signedMessage, machoFile.GetSignatureData().CmsData!!, isMacho = true)
    }
  }

  @ParameterizedTest
  @MethodSource("SignedMsiProvider")
  fun Msi_Test(signedResourceName: String) {
    val result = TestUtil.getTestByteChannel("msi", signedResourceName).use {
      val msiFile = MsiFile(it)
      val signatureData = msiFile.GetSignatureData()
      val signedMessage = SignedMessage.CreateInstance(signatureData)
      RecreateSignatureTest(signedMessage, msiFile.GetSignatureData().CmsData!!)
    }
  }

  /**
   * Tests, that we can recreate original signature
   */
  fun RecreateSignatureTest(
    signedMessage: SignedMessage,
    originalSignature: ByteArray,
    isMacho: Boolean = false
  ) {
    val signedData = signedMessage.SignedData
    val innerSignedData = signedData.signedData

    val contentInfo = signedData.contentInfo
    val initialSignedDataInfo = signedData.signedData.toASN1Primitive()

    val json = Json.encodeToString(Asn1PrimitiveSerializer, initialSignedDataInfo)
    val deserializedSignedDataInfo = Json.decodeFromString(Asn1PrimitiveSerializer, json)

    val signedDataInfo = SignedData.getInstance(deserializedSignedDataInfo)

    if (isMacho) {
      Assertions.assertEquals(
        true,
        compareBytes(
          innerSignedData.getEncoded("DER"),
          signedDataInfo.getEncoded("DER"),
          verbose = true
        )
      )

      val recreatedInfo = SignedData.getInstance(signedDataInfo).toContentInfo()

      Assertions.assertEquals(
        true,
        compareBytes(
          contentInfo.getEncoded("BER"),
          recreatedInfo.getEncoded("BER"),
          verbose = true
        )
      )

      val encoded = recreatedInfo.getEncoded("BER")

      Assertions.assertEquals(
        true,
        compareBytes(
          originalSignature,
          encoded,
          verbose = false
        )
      )
    } else {

      Assertions.assertEquals(
        true,
        compareBytes(
          innerSignedData.getEncoded("DER"),
          signedDataInfo.getEncoded("DER"),
          verbose = true
        )
      )

      val recreatedInfo = SignedData.getInstance(signedDataInfo).toContentInfo()

      Assertions.assertEquals(
        true,
        compareBytes(
          contentInfo.getEncoded("DER"),
          recreatedInfo.getEncoded("DER"),
          verbose = true
        )
      )

      val encoded = recreatedInfo.getEncoded("DER")

      Assertions.assertEquals(
        true,
        compareBytes(
          originalSignature,
          encoded,
          verbose = false
        )
      )
    }
  }

  companion object {
    private const val pe_01_signed = "ServiceModelRegUI.dll"

    private const val pe_02_signed = "self_signed_test.exe"

    private const val pe_03_signed = "shell32.dll"

    private const val pe_04_signed = "IntelAudioService.exe"

    private const val pe_05_signed = "libcrypto-1_1-x64.dll"

    private const val pe_06_signed = "libssl-1_1-x64.dll"

    private const val pe_07_signed = "JetBrains.dotUltimate.2021.3.EAP1D.Checked.web.exe"

    private const val pe_08_signed = "dotnet.exe"
    private const val pe_08_not_signed = "dotnet_no_sign.exe"

    @JvmStatic
    fun SignedMsiProvider(): Stream<Arguments> {
      return Stream.of(
        Arguments.of("2dac4b.msi"),
        Arguments.of("firefox.msi"),
        Arguments.of("sumatra.msi"),
      )
    }

    @JvmStatic
    fun SignedMachoProvider(): Stream<Arguments> {
      return Stream.of(
        Arguments.of("nosigned_resigned"),
        Arguments.of("env-wrapper.x64"),
        Arguments.of("libMonoSupportW.x64.dylib"),
        Arguments.of("cat"),
        Arguments.of("JetBrains.Profiler.PdbServer"),
        Arguments.of("fat.dylib_signed"),
        Arguments.of("libhostfxr.dylib")
      )
    }

    @JvmStatic
    fun SignedPEProvider(): Stream<Arguments> {
      return Stream.of(
        Arguments.of(
          pe_01_signed
        ),
        Arguments.of(
          pe_02_signed
        ),
        Arguments.of(
          pe_03_signed
        ),
        Arguments.of(
          pe_04_signed
        ),
        Arguments.of(
          pe_05_signed
        ),
        Arguments.of(
          pe_06_signed
        ),
        Arguments.of(
          pe_07_signed
        ),
        Arguments.of(
          pe_08_signed
        ),
        Arguments.of(
          "JetBrains.ReSharper.TestResources.dll"
        ),
        Arguments.of(
          "System.Security.Principal.Windows.dll"
        ),
      )
    }
  }
}

File: net/JetBrains.Serialization/tests/SignatureRecreationTests.cs
using JetBrains.FormatRipper.Compound;
using JetBrains.FormatRipper.MachO;
using JetBrains.FormatRipper.Pe;
using JetBrains.SignatureVerifier.Crypt;
using JetBrains.SignatureVerifier.Crypt.BC;
using Newtonsoft.Json;
using NUnit.Framework;
using Org.BouncyCastle.Asn1;
using Org.BouncyCastle.Asn1.Pkcs;
using ContentInfo = Org.BouncyCastle.Asn1.Cms.ContentInfo;
using SignedData = Org.BouncyCastle.Asn1.Cms.SignedData;

namespace JetBrains.Serialization.Tests;

public class SignatureRecreationTests
{
  // @formatter:off
  [TestCase("ServiceModelRegUI.dll")]
  [TestCase("shell32.dll")]
  [TestCase("IntelAudioService.exe")]
  [TestCase("libcrypto-1_1-x64.dll")]
  [TestCase("libssl-1_1-x64.dll")]
  [TestCase("JetBrains.dotUltimate.2021.3.EAP1D.Checked.web.exe")]
  // @formatter:on
  public Task PeRecreationTest(string resourceName)
  {
    var file = ResourceUtil.OpenRead(ResourceCategory.Pe, resourceName,
      stream => PeFile.Parse(stream, PeFile.Mode.SignatureData));

    var signedMessage = SignedMessage.CreateInstance(file.SignatureData);
    var originalContentInfo = signedMessage.SignedData.ContentInfo;
    RecreationTest(signedMessage.SignedData, file.SignatureData.CmsBlob!);

    return Task.CompletedTask;
  }

  // @formatter:off
  [TestCase("2dac4b.msi")]
  [TestCase("firefox.msi")]
  [TestCase("sumatra.msi")]
  // @formatter:on
  public Task MsiRecreationTest(string resourceName)
  {
    var file = ResourceUtil.OpenRead(ResourceCategory.Msi, resourceName, stream =>
    {
      Assert.IsTrue(CompoundFile.Is(stream));
      return CompoundFile.Parse(stream, CompoundFile.Mode.SignatureData);
    });

    var signedMessage = SignedMessage.CreateInstance(file.SignatureData);
    var originalContentInfo = signedMessage.SignedData.ContentInfo;
    RecreationTest(signedMessage.SignedData, file.SignatureData.CmsBlob!);

    return Task.CompletedTask;
  }

  private static MachOFile GetMachOFile(string resourceName) =>
    ResourceUtil.OpenRead(ResourceCategory.MachO, resourceName,
      stream => MachOFile.Parse(stream, MachOFile.Mode.SignatureData));

  // @formatter:off
  [TestCase("JetBrains.Profiler.PdbServer")]
  [TestCase("cat")]
  [TestCase("env-wrapper.x64")]
  [TestCase("libMonoSupportW.x64.dylib")]
  [TestCase("libhostfxr.dylib")]
  // @formatter:on
  public Task MachORecreationTest(string resourceName)
  {
    foreach (var section in GetMachOFile(resourceName).Sections)
    {
      var signedMessage = SignedMessage.CreateInstance(section.SignatureData);

      RecreationTest(signedMessage.SignedData, section.SignatureData.CmsBlob!, "BER");
    }

    return Task.CompletedTask;
  }

  public ContentInfo SignedDataToContentInfo(SignedData signedData, string encoding = "BER")
  {
    byte[] signedDataBytes = signedData.GetEncoded(encoding);

    Asn1InputStream inputStream = new Asn1InputStream(signedDataBytes);
    Asn1Object asn1Object = inputStream.ReadObject();

    return new ContentInfo(PkcsObjectIdentifiers.SignedData, asn1Object);
  }

  public void RecreationTest(CmsSignedData signedData, byte[] originalSignature, string encoding = "DER")
  {
    var innerSignedData = signedData.SignedData;
    var signedDataInfo = innerSignedData.ToAsn1Object();


    var json = JsonConvert.SerializeObject(signedDataInfo, new AsnJsonConverter());

    var recreated = JsonConvert.DeserializeObject<Asn1Object>(json, new AsnJsonConverter());

    var copy = SignedData.GetInstance(recreated);
    Assert.That(innerSignedData.GetEncoded("DER").SequenceEqual(copy.GetEncoded("DER")));

    var recreatedInfo = SignedDataToContentInfo(copy);
    Assert.That(signedData.ContentInfo.GetEncoded(encoding).SequenceEqual(recreatedInfo.GetEncoded(encoding)));

    var cropped = new ArraySegment<byte>(originalSignature, 0, recreatedInfo.GetEncoded(encoding).Length);
    Assert.That(cropped.ToArray().SequenceEqual(recreatedInfo.GetEncoded(encoding)));
  }
}